### Top tip: Cell types

Each cell has a 'type'. You can switch between types of cell with the dropdown in the top menu

- **Code cell**: is for python code. They are surrounded in grey boxes, and have line numbers.
- **Markdown (text) cell**: is for text (like this cell you're reading). They have white background and styling when done. Start lines with ### for a header, and surround words with ** for a highlight. **like this** 

(double click this cell to see the 'source'. Then run it to see the 'pretty' version)

# Who survived on Titanic?

In this lab we will use a famous leanring dataset about passangers of titanic. Titanic was a big passager boat which sunk on 15 April 1912. Lots of people died, but we have quite detailed information about passangers and whether they survived. On this lab we will investigate relationship between those variables (eg. wealth, age, gender, cabin number) and survival.

In [2]:
import pandas as pd

In [3]:
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
# this version of titanic dataset is taken from Kaggle
# https://www.kaggle.com/competitions/titanic/data?select=test.csv
passengers

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

# Task 1: Descriptive statistics

Describe the dataset in a set of bulletpoints (use Pandas to give you the answers). Try to backup each bulletpoint with separate pieces of python code.

Below are just some examples to get you started, but explore numbers and types of things (rows, columns). What are max min and mean of the numeric ones? Which columns have missing values and how many? etc.

For exaplanations of what columns mean see https://www.kaggle.com/competitions/titanic/data?select=test.csv

### A. What columns and of what type did you find? How many rows and columns?

In [4]:
passengers.shape

(891, 12)

In [5]:
passengers.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [6]:
passengers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [8]:
print("Mean", passengers.Age.mean())
# ...
# ...
# (you can add more cells wirh + button in the top menu)

Mean 29.69911764705882


**Answer A:**

there are 12 columns (+ index) and 887 rows. 

Columns have names: 'PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'

Name and Sex columns are an object (likely a string), Age and Fare are Floats, and Sibling/Parent rows are integers.

Age had a mean 29 years, and max and min are ... [you continue]

# Task 2. Answers basic questions: Select some rows, some columns and run some simple statistics

Explore ways to select only some wors and columns. eg. 

- just select name and age components of the data frame. Sort them by Age and get the names of 10 oldest passangers
- Pick the fares of pasangers who survived and did not survive and compare their ticket class (Pclass), Fare cost and sex. Is there a large difference in the means?

**Itentify 5 questions** like above and answer them with a simple sentence. Privide python code which will find you that answer. 

Call them 'A', 'B', 'C', ...

### But first some starting code. Read it, make sure you understand it. Feel free to modify it.

How to select only some rows and some columns?


Note about referencing: iloc[rows, columns] is used when referencing with numbers, and loc[rows, columns] is used when referencing by names. 

Also : mean 'all of them'

In [11]:
# first passanger, their data in column indexed 4
passengers.iloc[0,4]

'male'

In [12]:
passengers.iloc[0:5,[2,3,-1]] # first 5 passangers, only columns 2, 4 and the last one

Pclass                                               Name Embarked
0       3                            Braund, Mr. Owen Harris        S
1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...        C
2       3                             Heikkinen, Miss. Laina        S
3       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)        S
4       3                           Allen, Mr. William Henry        S

In [13]:
passengers.loc[10:20,['Age', 'Fare']] 
# passagners with indexes 10 to 20 passangers, only columns 'Age', 'Fare'

Age     Fare
10   4.0  16.7000
11  58.0  26.5500
12  20.0   8.0500
13  39.0  31.2750
14  14.0   7.8542
15  55.0  16.0000
16   2.0  29.1250
17   NaN  13.0000
18  31.0  18.0000
19   NaN   7.2250
20  35.0  26.0000

In [14]:
passengers.loc[:,['Name']] # All rows, column name

Name
0                              Braund, Mr. Owen Harris
1    Cumings, Mrs. John Bradley (Florence Briggs Th...
2                               Heikkinen, Miss. Laina
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                             Allen, Mr. William Henry
..                                                 ...
886                              Montvila, Rev. Juozas
887                       Graham, Miss. Margaret Edith
888           Johnston, Miss. Catherine Helen "Carrie"
889                              Behr, Mr. Karl Howell
890                                Dooley, Mr. Patrick

[891 rows x 1 columns]

In [15]:
# finally, you can combine them:
# get first 5 people, and their Names
passengers.iloc[0:5,:].loc[:,'Name']

0                              Braund, Mr. Owen Harris
1    Cumings, Mrs. John Bradley (Florence Briggs Th...
2                               Heikkinen, Miss. Laina
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                             Allen, Mr. William Henry
Name: Name, dtype: object

To pick first or last n items you can also use .head(n) or tail(n).

And if picking just one column (and it does not have spaces in the name), you can use .ColumnName 

In [16]:
passengers.head(5).Name

0                              Braund, Mr. Owen Harris
1    Cumings, Mrs. John Bradley (Florence Briggs Th...
2                               Heikkinen, Miss. Laina
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                             Allen, Mr. William Henry
Name: Name, dtype: object

Ok, let;s find our 5 questions and answer them (call them A,B,C,D and E)

### A: Names of oldest passangers. What can we tell about their Survival rate when compared with youngest ones?

here's a start on the first one

In [17]:
# just select name and age components of the data frame. 
dfA2 = passengers[['Age', 'Name', 'Pclass', 'Survived']].copy()

# Sort them by Age 
dfA2.sort_values(by='Age', ascending=False, inplace=True)

# and get the names of 10 oldest passangers
oldest_10 = dfA2.iloc[0:10]
oldest_10

Age                                  Name  Pclass  Survived
630  80.0  Barkworth, Mr. Algernon Henry Wilson       1         1
851  74.0                   Svensson, Mr. Johan       3         0
493  71.0               Artagaveytia, Mr. Ramon       1         0
96   71.0             Goldschmidt, Mr. George B       1         0
116  70.5                  Connors, Mr. Patrick       3         0
672  70.0           Mitchell, Mr. Henry Michael       2         0
745  70.0          Crosby, Capt. Edward Gifford       1         0
33   66.0                 Wheadon, Mr. Edward H       2         0
54   65.0        Ostby, Mr. Engelhart Cornelius       1         0
280  65.0                      Duane, Mr. Frank       3         0

In [18]:
# note: and if you wanted to get names of all people over 70
print(dfA2[dfA2.Age > 70].loc[:, ['Name','Age']])

                                     Name   Age
630  Barkworth, Mr. Algernon Henry Wilson  80.0
851                   Svensson, Mr. Johan  74.0
493               Artagaveytia, Mr. Ramon  71.0
96              Goldschmidt, Mr. George B  71.0
116                  Connors, Mr. Patrick  70.5


In [19]:
# Or get names of all people over 60 who survived
print(dfA2[(dfA2.Age > 60)&(dfA2.Survived == 1)].loc[:, ['Name','Age']])

                                          Name   Age
630       Barkworth, Mr. Algernon Henry Wilson  80.0
275          Andrews, Miss. Kornelia Theodosia  63.0
483                     Turkula, Mrs. (Hedwig)  63.0
570                         Harris, Mr. George  62.0
829  Stone, Mrs. George Nelson (Martha Evelyn)  62.0


In [20]:
print('oldest 10 survival', dfA2.iloc[0:10,:].loc[:, 'Survived'].mean())
# could also be
print('oldest 10 survival', dfA2.iloc[0:10,:].Survived.mean())
print('oldest 10 survival', dfA2.head(10).Survived.mean())

oldest 10 survival 0.1
oldest 10 survival 0.1
oldest 10 survival 0.1


In [21]:
print('youngest 10 survival', dfA2.tail(10).Survived.mean())

youngest 10 survival 0.2


In [22]:
print('average all survival', dfA2.Survived.mean())

average all survival 0.3838383838383838


Answer: Name of the oldest passanger (80) was Mr. Algernon Henry Wilson Barkworth. Survival rate of 10 oldest passangers was 10%, and of the youngest 10 passagners 90%. In context, the average survival rate was 38%.

### B: What's the impact of Passanger Class and Sex on the suvival?

In [23]:
# your code here (add more cells with '+' in the top menu)

You answer here:

### C: Your question? Select a subset, and look at it's qualities

In [24]:
# your code here (add more cells with '+' in the top menu)

You answer here:

### D: Your question?

In [25]:
# your code here (add more cells with '+' in the top menu)

You answer here:

### E: Your question?

In [26]:
# your code here (add more cells with '+' in the top menu)

You answer here:

# Task 3: Cleaning up and improving the dataset:

A: Identify missing values and decide what to do with them. You can remove whole rows with missing things, or you can fill them with averages. Or averages for a particular subgroup.

B: Rename the column names so that they follow the same naming convention (eg. no spaces, CamelCase or snake_case). 

C: Can you create new columns which use available information? eg. Can you create a column TravelingAlone based on the values in siblings/partners and parents/children ?

### A: Looking at missing values

In [27]:
# btw. re-run this cell to reset yoru dataset it its original form
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')


In [28]:
# all data into binary: is missing or not:
passengers.isna()

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0          False     False   False  False  False  False  False  False   False   
1          False     False   False  False  False  False  False  False   False   
2          False     False   False  False  False  False  False  False   False   
3          False     False   False  False  False  False  False  False   False   
4          False     False   False  False  False  False  False  False   False   
..           ...       ...     ...    ...    ...    ...    ...    ...     ...   
886        False     False   False  False  False  False  False  False   False   
887        False     False   False  False  False  False  False  False   False   
888        False     False   False  False  False   True  False  False   False   
889        False     False   False  False  False  False  False  False   False   
890        False     False   False  False  False  False  False  False   False   

      Fare  Cabin  Embarked  
0    False   True     False  
1    False  False     False  
2    False   True     False  
3    False  False     False  
4    False   True     False  
..     ...    ...       ...  
886  False   True     False  
887  False  False     False  
888  False   True     False  
889  False  False     False  
890  False   True     False  

[891 rows x 12 columns]

In [29]:
# by columns, are any items True?
passengers.isna().any()

PassengerId    False
Survived       False
Pclass         False
Name           False
Sex            False
Age             True
SibSp          False
Parch          False
Ticket         False
Fare           False
Cabin           True
Embarked        True
dtype: bool

In [30]:
# how many passagners would you loose if you passengers.dropna(axis=0) ?
# you can use .shape()
pass_no_na = passengers.dropna(axis=0)
pass_no_na

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
10            11         1       3   
11            12         1       1   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                           Carlsson, Mr. Frans Olof    male  33.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
1        0  PC 17599  71.2833          C85        C  
3        0    113803  53.1000         C123        S  
6        0     17463  51.8625          E46        S  
10       1   PP 9549  16.7000           G6        S  
11       0    113783  26.5500         C103        S  
..     ...       ...      ...          ...      ...  
871      1     11751  52.5542          D35        S  
872      0       695   5.0000  B51 B53 B55        S  
879      1     11767  83.1583          C50        C  
887      0    112053  30.0000          B42        S  
889      0    111369  30.0000         C148        C  

[183 rows x 12 columns]

In [31]:
# You could replace all missing values with a set value, 
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
pass_no_na = passengers.fillna(0) 
# but what unexpected effects it could have?
pass_no_na

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   0.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500     0        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250     0        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500     0        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000     0        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500     0        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500     0        Q  

[891 rows x 12 columns]

In [32]:
# or you can replace them with a most common value,
# or other statistic of that column eg.mean, max
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
passengers.fillna(passengers.mean(numeric_only=True)) 

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex        Age  \
0                              Braund, Mr. Owen Harris    male  22.000000   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.000000   
2                               Heikkinen, Miss. Laina  female  26.000000   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.000000   
4                             Allen, Mr. William Henry    male  35.000000   
..                                                 ...     ...        ...   
886                              Montvila, Rev. Juozas    male  27.000000   
887                       Graham, Miss. Margaret Edith  female  19.000000   
888           Johnston, Miss. Catherine Helen "Carrie"  female  29.699118   
889                              Behr, Mr. Karl Howell    male  26.000000   
890                                Dooley, Mr. Patrick    male  32.000000   

     SibSp  Parch            Ticket     Fare Cabin Embarked  
0        1      0         A/5 21171   7.2500   NaN        S  
1        1      0          PC 17599  71.2833   C85        C  
2        0      0  STON/O2. 3101282   7.9250   NaN        S  
3        1      0            113803  53.1000  C123        S  
4        0      0            373450   8.0500   NaN        S  
..     ...    ...               ...      ...   ...      ...  
886      0      0            211536  13.0000   NaN        S  
887      0      0            112053  30.0000   B42        S  
888      1      2        W./C. 6607  23.4500   NaN        S  
889      0      0            111369  30.0000  C148        C  
890      0      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [33]:
# for non-numeric values one of the oprtions is to use the most common (mode) value.
# since there might be mady most common values, you'll have to specify that you want the first one
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
passengers.fillna(passengers.mode().iloc[0]) 
# notice what happens with the Cabin

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  24.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare    Cabin Embarked  
0        0         A/5 21171   7.2500  B96 B98        S  
1        0          PC 17599  71.2833      C85        C  
2        0  STON/O2. 3101282   7.9250  B96 B98        S  
3        0            113803  53.1000     C123        S  
4        0            373450   8.0500  B96 B98        S  
..     ...               ...      ...      ...      ...  
886      0            211536  13.0000  B96 B98        S  
887      0            112053  30.0000      B42        S  
888      2        W./C. 6607  23.4500  B96 B98        S  
889      0            111369  30.0000     C148        C  
890      0            370376   7.7500  B96 B98        Q  

[891 rows x 12 columns]

In [34]:
# you could also sopecify individual columns and values
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
passengers.Cabin.fillna("NoCabin", inplace=True) 
passengers

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare    Cabin Embarked  
0        0         A/5 21171   7.2500  NoCabin        S  
1        0          PC 17599  71.2833      C85        C  
2        0  STON/O2. 3101282   7.9250  NoCabin        S  
3        0            113803  53.1000     C123        S  
4        0            373450   8.0500  NoCabin        S  
..     ...               ...      ...      ...      ...  
886      0            211536  13.0000  NoCabin        S  
887      0            112053  30.0000      B42        S  
888      2        W./C. 6607  23.4500  NoCabin        S  
889      0            111369  30.0000     C148        C  
890      0            370376   7.7500  NoCabin        Q  

[891 rows x 12 columns]

In [35]:
# Your task, decide what to do with the missing values in the origibnal dataset:
# Once your_dataset.isna().any() returns all False, there are no more missing values

### B: Duplicates. 

In this dataset there are no dulplicates. So here's a small exercise. Let's duplicate the first person so they appear 3 times. And let's change the age of one of those entries. So we're pretending first passanger got accidentally added 3 times, including once with a slightly different data for age.

In [36]:
# att two more copies of first row
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
with_duplicates = passengers
with_duplicates = pd.concat((passengers.iloc[0:1], passengers.iloc[0:1], passengers))
with_duplicates

PassengerId  Survived  Pclass  \
0              1         0       3   
0              1         0       3   
0              1         0       3   
1              2         1       1   
2              3         1       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
0                              Braund, Mr. Owen Harris    male  22.0      1   
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
0        0         A/5 21171   7.2500   NaN        S  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[893 rows x 12 columns]

In [37]:
# now let's fix the index
with_duplicates.reset_index(drop=True, inplace=True) # drop completely forgets previous index
with_duplicates

PassengerId  Survived  Pclass  \
0              1         0       3   
1              1         0       3   
2              1         0       3   
3              2         1       1   
4              3         1       3   
..           ...       ...     ...   
888          887         0       2   
889          888         1       1   
890          889         0       3   
891          890         1       1   
892          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1                              Braund, Mr. Owen Harris    male  22.0      1   
2                              Braund, Mr. Owen Harris    male  22.0      1   
3    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
4                               Heikkinen, Miss. Laina  female  26.0      0   
..                                                 ...     ...   ...    ...   
888                              Montvila, Rev. Juozas    male  27.0      0   
889                       Graham, Miss. Margaret Edith  female  19.0      0   
890           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
891                              Behr, Mr. Karl Howell    male  26.0      0   
892                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0         A/5 21171   7.2500   NaN        S  
2        0         A/5 21171   7.2500   NaN        S  
3        0          PC 17599  71.2833   C85        C  
4        0  STON/O2. 3101282   7.9250   NaN        S  
..     ...               ...      ...   ...      ...  
888      0            211536  13.0000   NaN        S  
889      0            112053  30.0000   B42        S  
890      2        W./C. 6607  23.4500   NaN        S  
891      0            111369  30.0000  C148        C  
892      0            370376   7.7500   NaN        Q  

[893 rows x 12 columns]

In [38]:
# change one of the duplicates to have another age:
with_duplicates.loc[1,'Age'] = 23 

# Note: as well as loc[] and iloc[], which access many items,
# you can use at[] and iat[] to access just one:
# with_duplicates.at[1,'Age'] = 23

with_duplicates

PassengerId  Survived  Pclass  \
0              1         0       3   
1              1         0       3   
2              1         0       3   
3              2         1       1   
4              3         1       3   
..           ...       ...     ...   
888          887         0       2   
889          888         1       1   
890          889         0       3   
891          890         1       1   
892          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1                              Braund, Mr. Owen Harris    male  23.0      1   
2                              Braund, Mr. Owen Harris    male  22.0      1   
3    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
4                               Heikkinen, Miss. Laina  female  26.0      0   
..                                                 ...     ...   ...    ...   
888                              Montvila, Rev. Juozas    male  27.0      0   
889                       Graham, Miss. Margaret Edith  female  19.0      0   
890           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
891                              Behr, Mr. Karl Howell    male  26.0      0   
892                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0         A/5 21171   7.2500   NaN        S  
2        0         A/5 21171   7.2500   NaN        S  
3        0          PC 17599  71.2833   C85        C  
4        0  STON/O2. 3101282   7.9250   NaN        S  
..     ...               ...      ...   ...      ...  
888      0            211536  13.0000   NaN        S  
889      0            112053  30.0000   B42        S  
890      2        W./C. 6607  23.4500   NaN        S  
891      0            111369  30.0000  C148        C  
892      0            370376   7.7500   NaN        Q  

[893 rows x 12 columns]

Now let's check for duplicates, and remove them. You might want to define a duplicate as when some unique element (or a combination of unique elements) are the same.

In [39]:
with_duplicates.duplicated()
# here you should see that the person at index 2 is a duplicate (perfectly identical)
# you could also decide to keep last (not first) with  duplicated(keep='last')

0      False
1      False
2       True
3      False
4      False
       ...  
888    False
889    False
890    False
891    False
892    False
Length: 893, dtype: bool

In [40]:
no_duplicates1 = with_duplicates.drop_duplicates()
no_duplicates1
# this would remove all completely-dentical duplicates

PassengerId  Survived  Pclass  \
0              1         0       3   
1              1         0       3   
3              2         1       1   
4              3         1       3   
5              4         1       1   
..           ...       ...     ...   
888          887         0       2   
889          888         1       1   
890          889         0       3   
891          890         1       1   
892          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1                              Braund, Mr. Owen Harris    male  23.0      1   
3    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
4                               Heikkinen, Miss. Laina  female  26.0      0   
5         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
..                                                 ...     ...   ...    ...   
888                              Montvila, Rev. Juozas    male  27.0      0   
889                       Graham, Miss. Margaret Edith  female  19.0      0   
890           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
891                              Behr, Mr. Karl Howell    male  26.0      0   
892                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0         A/5 21171   7.2500   NaN        S  
3        0          PC 17599  71.2833   C85        C  
4        0  STON/O2. 3101282   7.9250   NaN        S  
5        0            113803  53.1000  C123        S  
..     ...               ...      ...   ...      ...  
888      0            211536  13.0000   NaN        S  
889      0            112053  30.0000   B42        S  
890      2        W./C. 6607  23.4500   NaN        S  
891      0            111369  30.0000  C148        C  
892      0            370376   7.7500   NaN        Q  

[892 rows x 12 columns]

In [41]:
# this would remove all duplicates, identical only in those columns
no_duplicates2 = with_duplicates.drop_duplicates(subset=['Name'])
no_duplicates2
# notice we're down to one duplicated first passanger

PassengerId  Survived  Pclass  \
0              1         0       3   
3              2         1       1   
4              3         1       3   
5              4         1       1   
6              5         0       3   
..           ...       ...     ...   
888          887         0       2   
889          888         1       1   
890          889         0       3   
891          890         1       1   
892          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
3    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
4                               Heikkinen, Miss. Laina  female  26.0      0   
5         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
888                              Montvila, Rev. Juozas    male  27.0      0   
889                       Graham, Miss. Margaret Edith  female  19.0      0   
890           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
891                              Behr, Mr. Karl Howell    male  26.0      0   
892                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
3        0          PC 17599  71.2833   C85        C  
4        0  STON/O2. 3101282   7.9250   NaN        S  
5        0            113803  53.1000  C123        S  
6        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
888      0            211536  13.0000   NaN        S  
889      0            112053  30.0000   B42        S  
890      2        W./C. 6607  23.4500   NaN        S  
891      0            111369  30.0000  C148        C  
892      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [42]:
# so now there are no duplicates
no_duplicates2.duplicated()


0      False
3      False
4      False
5      False
6      False
       ...  
888    False
889    False
890    False
891    False
892    False
Length: 891, dtype: bool

In [43]:
# and we can re-index
no_duplicates2.reset_index(drop=True, inplace=True)
no_duplicates2

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

### C: Improving Column names

Look at the notes if you're unsure how to do it. You're looking for rename(). Aim for no spaces or special characters in column names, so that you can use . notation, like .Age

One improvementc could be making those columns more human readable: SibSp, Parch. See link to Kaggle above for more insights or what they mean/

### D: Categorical variables (in R called Vectors, in Java called Enums)

Some values are presented as numbers (Pclass), but really, they are categories. Some are presented as strings (sex), but with some careful consideration and kind design can be analysed as categories too (notice this very old dataset is using Sex, not gender*).

**Note on usage of Gender and Sex in datsets: read up on this after the lab**

*In all modern datasets Gender is represented as much more complex datastructure than just True/False binary (this was common in early days of databases, where gender was a True/False value 'is_male'). Later it was a category (with 2 or 3 choices eg. 'male', 'female', 'prefer_not_to_say'). 

Luckilly, to reflect beauty of human experience there are some very interesting research projects on how to design inclusive forms and datsets. This Scottish Government oriject involved data scientists and people with various experiences of gender. Do some reading on the subject https://www.gov.scot/publications/reviewing-design-methods-make-more-sensitive-gender/pages/1/ and notice it does not provide 'one answer fits all' but rather encourages empathy on project-by-project basis. Also form designers together with LGBTQI+ organisations are constanly advancing this field https://uxdesign.cc/designing-forms-for-gender-diversity-and-inclusion-d8194cf1f51  

So let's not look at the Sex field, but rather Pclass.

Let's first have a look at exploring Passanger class: What passanger classess were there on the titanic (on british trains we still have 1st and 2nd class carriages

In the dataset they are represented 

In [44]:
passengers.Pclass

0      3
1      1
2      3
3      1
4      3
      ..
886    2
887    1
888    3
889    1
890    3
Name: Pclass, Length: 891, dtype: int64

In [45]:
passengers.Pclass.value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [46]:
# are there any NA ?
passengers.Pclass.isna().value_counts()

False    891
Name: Pclass, dtype: int64

In [47]:
passengers['TicketClass'] = pd.Categorical(passengers.Pclass)
passengers

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked TicketClass  
0        0         A/5 21171   7.2500   NaN        S           3  
1        0          PC 17599  71.2833   C85        C           1  
2        0  STON/O2. 3101282   7.9250   NaN        S           3  
3        0            113803  53.1000  C123        S           1  
4        0            373450   8.0500   NaN        S           3  
..     ...               ...      ...   ...      ...         ...  
886      0            211536  13.0000   NaN        S           2  
887      0            112053  30.0000   B42        S           1  
888      2        W./C. 6607  23.4500   NaN        S           3  
889      0            111369  30.0000  C148        C           1  
890      0            370376   7.7500   NaN        Q           3  

[891 rows x 13 columns]

In [48]:
# Categorical data has number of advantages in generating statistics etc
# but also make our datasts much smaller (in memory) and faster (irn processing)

passengers.memory_usage()
# notice ticketclass stores the same data as Pclass, but uses 15% of memory.

Index           128
PassengerId    7128
Survived       7128
Pclass         7128
Name           7128
Sex            7128
Age            7128
SibSp          7128
Parch          7128
Ticket         7128
Fare           7128
Cabin          7128
Embarked       7128
TicketClass    1023
dtype: int64

In [49]:
# Additionally you can specify the order of categories, so that you 
# could perform comparisons

# introduce order
passengers['TicketClass'] = pd.Categorical(passengers.Pclass, 
                                           categories=[3,2,1], # here specify order
                                           ordered=True
                                          )
# optionally re-name categories with cat.rename_categories
mapping = {1: 'Luxury', 2: 'Economy', 3: 'LowerDeck'}
passengers['TicketClass'] = passengers['TicketClass'].cat.rename_categories(mapping)
passengers['TicketClass'] # notice printout shows order of categories

0      LowerDeck
1         Luxury
2      LowerDeck
3         Luxury
4      LowerDeck
         ...    
886      Economy
887       Luxury
888    LowerDeck
889       Luxury
890    LowerDeck
Name: TicketClass, Length: 891, dtype: category
Categories (3, object): ['LowerDeck' < 'Economy' < 'Luxury']

In [50]:
# so that you can now eg choose all TicketClass LOWER THAN a number:
passengers[passengers['TicketClass'] < 'Luxury']

PassengerId  Survived  Pclass                                      Name  \
0              1         0       3                   Braund, Mr. Owen Harris   
2              3         1       3                    Heikkinen, Miss. Laina   
4              5         0       3                  Allen, Mr. William Henry   
5              6         0       3                          Moran, Mr. James   
7              8         0       3            Palsson, Master. Gosta Leonard   
..           ...       ...     ...                                       ...   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S   
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
4      male  35.0      0      0            373450   8.0500   NaN        S   
5      male   NaN      0      0            330877   8.4583   NaN        Q   
7      male   2.0      3      1            349909  21.0750   NaN        S   
..      ...   ...    ...    ...               ...      ...   ...      ...   
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S   
885  female  39.0      0      5            382652  29.1250   NaN        Q   
886    male  27.0      0      0            211536  13.0000   NaN        S   
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S   
890    male  32.0      0      0            370376   7.7500   NaN        Q   

    TicketClass  
0     LowerDeck  
2     LowerDeck  
4     LowerDeck  
5     LowerDeck  
7     LowerDeck  
..          ...  
884   LowerDeck  
885   LowerDeck  
886     Economy  
888   LowerDeck  
890   LowerDeck  

[675 rows x 13 columns]

In [51]:
# what other columns could be categories. Try them!

### E: Creating more columns

Look at the notes if you're unsure how to do it. What could you use: age groups? In one og the notebooks you were shown how to group different ranges of values into groups. Maybe you could use that technique on age?

Or maybe you could combine columns abotu partners and dependants (SibSp,Parch) into a column TravelsAlone?

How do Survival stats look like for different values of the new columns?